### PR TITLE
add dynamic list of todos solution

### PR DESCRIPTION
**Pull request by opiquem**
[DEMO LINK](https://maksym-bobryk.github.io/react_dynamic-list-of-todos/)

### CODE REVIEW

File: src/components/TodoFilter/TodoFilter.tsx [hunk 1]
```diff
@@ -1,34 +1,72 @@
...
-      </span>
-    </p>
-  </form>
-);
+        <span className="icon is-left">
+          <i className="fas fa-magnifying-glass" />
+        </span>
+        {query && (
+          <span className="icon is-right" style={{ pointerEvents: 'all' }}>
+            {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
```
Comment (annholovko) [inline src/components/TodoFilter/TodoFilter.tsx:59]: I don't know whether this is needed and whether it won't be a mistake , but u can add aria-label to your button and delete this comment

File: src/App.tsx [hunk 2]
```diff
@@ -1,14 +1,51 @@
...
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [query, setQuery] = useState<string>('');
+  const [status, setStatus] = useState<TodoStatus>(TodoStatus.All);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+      });
```
Comment (annholovko) [inline src/App.tsx:26]: u can add there `catch` to catch some errors😁

File: src/components/TodoModal/TodoModal.tsx [hunk 3]
```diff
@@ -15,30 +37,33 @@ export const TodoModal: React.FC = () => {
               className="modal-card-title has-text-weight-medium"
               data-cy="modal-header"
             >
-              Todo #2
+              {`Todo #${selectedTodo.id}`}
             </div>
 
             {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
```
Comment (annholovko) [inline src/components/TodoModal/TodoModal.tsx:43]: I don't know whether this is needed and whether it won't be a mistake , but u can add aria-label to your button and delete this comment

File: src/components/TodoModal/TodoModal.tsx [hunk 4]
```diff
@@ -15,30 +37,33 @@ export const TodoModal: React.FC = () => {
...
           </header>
 
           <div className="modal-card-body">
             <p className="block" data-cy="modal-title">
-              quis ut nam facilis et officia qui
+              {selectedTodo.title}
             </p>
 
             <p className="block" data-cy="modal-user">
               {/* <strong className="has-text-success">Done</strong> */}
```
Comment (annholovko) [inline src/components/TodoModal/TodoModal.tsx:58]:
```suggestion

```

File: src/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  //   filteredTodos = filteredTodos.filter(todo => (
+  //     todo.title.toLowerCase().includes(loweredQuery)));
+  // }
+
+  // switch (status) {
+  //   case TodoStatus.Active:
+  //     return filteredTodos.filter()
+  //   default:
+  //     break;
+  // }
```
Comment (annholovko) [inline src/helpers.ts:23]: it will be better to remove this comments if your don't use this code

File: src/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        break;
+
+      default:
+        todoFilterStatus = true;
+        break;
+    }
+
+    return loweredTodoTitle.includes(loweredQuery) && todoFilterStatus;
+  });
+};
```
Comment (annholovko) [inline src/helpers.ts:46]: good desicion to create helper function for this🔥

File: src/App.tsx [hunk 7]
```diff
@@ -1,14 +1,51 @@
...
+  }, []);
```
Comment (okosohor) [inline src/App.tsx:27]: `
useEffect(() => {
    getTodos()
      .then(setTodos);
  }, []);
`

File: src/components/TodoModal/TodoModal.tsx [hunk 8]
```diff
@@ -1,12 +1,34 @@
...
+};
+
+export const TodoModal: React.FC<Props> = ({ selectedTodo, onClearForm }) => {
+  const [user, setUser] = useState<User | null>(null);
+
+  useEffect(() => {
+    getUser(selectedTodo.userId)
+      .then(userFromServer => {
+        setUser(userFromServer);
+      });
```
Comment (annholovko) [inline src/components/TodoModal/TodoModal.tsx:19]: same here: u can use `catch` to catch errors

**Review by annholovko — CHANGES_REQUESTED** [review]
GJ🔥
almost done, left some comments

File: src/helpers.ts (hunk 5 above)
Comment (okosohor) [inline src/helpers.ts:23]: remove it

**Review by okosohor — APPROVED** [review]

File: src/App.tsx [hunk 9]
```diff
@@ -17,18 +57,36 @@ export const App: React.FC = () => {
...
+                status={status}
+                onInputQuery={inputQuery}
+                onSelectStatus={selectStatus}
+              />
             </div>
 
             <div className="block">
-              <Loader />
-              <TodoList />
+              {todos.length === 0
```
Comment (33Orange) [inline src/App.tsx:69]: Let's add some boolean flag `isLoading` and use it here.
You can create boolean useState with initial value `true` and change it to `false` after you get a response from the server

File: src/components/TodoModal/TodoModal.tsx [hunk 10]
```diff
@@ -15,30 +40,33 @@ export const TodoModal: React.FC = () => {
               className="modal-card-title has-text-weight-medium"
               data-cy="modal-header"
             >
-              Todo #2
+              {`Todo #${selectedTodo.id}`}
             </div>
 
             {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
```
Comment (33Orange) [inline src/components/TodoModal/TodoModal.tsx:46]: Let's add aria-label to the button and remove `eslint-disable`

File: src/components/TodoModal/TodoModal.tsx [hunk 11]
```diff
@@ -15,30 +40,33 @@ export const TodoModal: React.FC = () => {
...
           </header>
 
           <div className="modal-card-body">
             <p className="block" data-cy="modal-title">
-              quis ut nam facilis et officia qui
+              {selectedTodo.title}
             </p>
 
             <p className="block" data-cy="modal-user">
               {/* <strong className="has-text-success">Done</strong> */}
```
Comment (33Orange) [inline src/components/TodoModal/TodoModal.tsx:61]: Remove redundant comments

File: src/components/TodoInfo/TodoInfo.tsx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+              'fa-eye-slash': isTodoSelected,
+            })}
+            />
+          </span>
+        </button>
+      </td>
+    </tr>
+  );
+};
+// cn({'far fa-eye', 'slash': isTodoSelected})
```
Comment (33Orange) [inline src/components/TodoInfo/TodoInfo.tsx:55]: Remove redundant comments

File: src/components/TodoInfo/TodoInfo.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        >
+          {todo.title}
+        </p>
+      </td>
+      <td className="has-text-right is-vcentered">
+        <button
+          data-cy="selectButton"
+          className="button"
+          type="button"
+          onClick={() => onSelectTodo(todo)}
```
Comment (33Orange) [inline src/components/TodoInfo/TodoInfo.tsx:41]: Crete separate function

File: src/components/TodoFilter/TodoFilter.tsx (hunk 1 above)
Comment (33Orange) [inline src/components/TodoFilter/TodoFilter.tsx:59]: Same here about aria-label

**Review by 33Orange — CHANGES_REQUESTED** [review]
Almost done

File: src/components/TodoInfo/TodoInfo.tsx [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  selectedTodo: Todo | null,
+};
+
+export const TodoInfo: React.FC<Props> = ({
+  todo,
+  onSelectTodo,
+  selectedTodo,
+}) => {
+  const isTodoSelected = todo.id === selectedTodo?.id;
+  const selectTodo = () => (
```
Comment (33Orange) [inline src/components/TodoInfo/TodoInfo.tsx:16]: We create function names with the verb `handle` or `on`

 ```suggestion
  const onClickSelectTodo = () => (
```

**Review by 33Orange — APPROVED** [review]
Well done

**Pull request closed without merging**